compiler/pipes/transform-to-smart-instanceof.cpp

```diff
@@ -14,14 +33,24 @@ bool TransformToSmartInstanceof::user_recursion(VertexPtr v) {
     return false;
   }
 
-  if (!infer_class_of_expr(current_function, instance_var).try_as_class()) {
-    kphp_error(false, "left operand of `instanceof` should be instance of class");
-    return false;
-  }
+  auto type_of_lhs = infer_class_of_expr(current_function, instance_var).try_as_class();
+  kphp_error_act(type_of_lhs, "left operand of `instanceof` should be instance of class", return false);
 
   run_function_pass(if_vertex->cond(), this);
 
-  add_tmp_var_with_instance_cast(instance_var, condition->rhs(), if_vertex->true_cmd_ref());
+  auto name_of_rhs_class_vertex = condition->rhs().as<op_string>();
+  const std::string &name_of_rhs_class = name_of_rhs_class_vertex->str_val;
+  ClassPtr rhs_class = G->get_class(name_of_rhs_class);
+  kphp_error_act(rhs_class, fmt_format("can't find right operand of `instanceof`: {}", name_of_rhs_class), return false);
+
+  if (!rhs_class->is_parent_of(type_of_lhs) && !type_of_lhs->is_parent_of(rhs_class)) {
+    rhs_class = get_first_common_child(rhs_class, type_of_lhs);
```

Comment:
The following example works incorrect on KPHP :(
```php
<?php

interface A {
  public function f_a();
}

interface B {
  public function f_b();
}

class AB implements A, B {
  public function f_a() { echo "AB::f_a\n"; }
  public function f_b() { echo "AB::f_b\n"; }
}

class AB2 implements A, B {
    public function f_a() { echo "AB2::f_a\n"; }
    public function f_b() { echo "AB2::f_b\n"; }
}

function demo(A $a) {
  if ($a instanceof B) {
    $a->f_b();
    $a->f_a();
  }
}

demo(new AB());
demo(new AB2());
```

PHP out:
```
AB::f_b
AB::f_a
AB2::f_b
AB2::f_a
```

KPHP out:
```
AB::f_b
AB::f_a
AB::f_b
AB::f_a
```

Reply:
Yeap, you are right this solution doesn't work at all. I'd close this PR and fix it later with a completely different approach.